File: python/lidy_test.py
import re

import specimen

interpolation_regex = r"\$\{([a-zA-Z0-9_]+)\}"


def template_read_entry(input: dict[str, str], key: str):
    value = input.get(key)
    template_value = input.get(f"{key}Template")
    if value is not None and template_value is not None:
        raise ValueError(
            f"Found both '{key}' and '{key}Template'. Only one must be specified."
        )
    if value is None and template_value is None:
        return None
    if value is not None:
        return value

    def handle_replace(match) -> str:
        name = match[2:-1]
        value = input.get(name)
        if value is None:
            raise ValueError(
                f"the template interpolation key '{name}' was not found in the input"
            )
        return value

    return re.sub(interpolation_regex, handle_replace, template_value)


@specimen.run(
    specimen.read_local_file("../../testdata/collection/listOf.spec.yaml"),
    specimen.read_local_file("../../testdata/collection/map.spec.yaml"),
    specimen.read_local_file("../../testdata/collection/mapOf.spec.yaml"),
    specimen.read_local_file("../../testdata/collection/merge.spec.yaml"),
    specimen.read_local_file("../../testdata/collection/min_max_nb.spec.yaml"),
    specimen.read_local_file("../../testdata/collection/tuple.spec.yaml"),
    specimen.read_local_file("../../testdata/combinator/oneOf.spec.yaml"),
    specimen.read_local_file("../../testdata/scalar/in.spec.yaml"),
    specimen.read_local_file("../../testdata/scalar/range.spec.yaml"),
    specimen.read_local_file("../../testdata/scalar/regexp.spec.yaml"),
    specimen.read_local_file("../../testdata/scalarType/scalar.spec.yaml"),
    specimen.read_local_file("../../testdata/schema/document.spec.yaml"),
    specimen.read_local_file("../../testdata/schema/expression.spec.yaml"),
    specimen.read_local_file("../../testdata/schema/mergeChecker.spec.yaml"),
    specimen.read_local_file("../../testdata/schema/regex.spec.yaml"),
    specimen.read_local_file("../../testdata/yaml/yaml.spec.yaml"),
)
def test_handler(
    s, name: str, box="content", errorContains: str | None = None, **kwargs
):
    text = template_read_entry(kwargs, "text")
    if text is None:
        s.fail("The 'text' entry is required")
    expression = template_read_entry(kwargs, "expression")
    schema = template_read_entry(kwargs, "schema")

    if box == "content":
        if expression is not None and schema is not None:
            s.fail("'expression' and 'schema' cannot be specified together")
        if expression is None and schema is None:
            s.fail("one of 'expression' and 'schema' must be specified")
        if expression is not None:
            schema = f"main:\n  {expression.replace('\n', '\n  ')}"
    else:
        word = ""
        if expression is not None:
            word = "expression"
        elif schema is not None:
            word = "schema"
        if word:
            s.fail(f"box {box} should not receive any {word}")

    outcome = ""
    if name.startswith("accept"):
        outcome = "accept"
    elif name.startswith("reject"):
        outcome = "reject"
    if outcome == "":
        s.fail(
            f"the 'name' entry should begin by 'accept' or 'reject', but it is: {name}"
        )

    error = None
    if box == "content":
        if schema == "":
            s.fail("the schema cannot be empty")
        try:
            parser = lidy.make_parser(
                lidy.File(name="<schema>.yaml", content=schema),
                dict(),
            )
        except Exception as e:
            error = e
        if error:
            s.abort(f"error in schema: {error}")
        try:
            parser.parse(lidy.File(name="<content>.yaml", content=text))
        except Exception as e:
            error = e
    else:
        if box == "lidySchemaExpression":
            schema = f"main:\n  {text.replace('\n', '\n  ')}"
        elif box == "lidySchemaDocument":
            schema = text
        elif box == "lidySchemaRegexChecker":
            schema = f"main:\n  _regex: '{text}'"
        else:
            s.fail(f"unknown test box: {box}")
        try:
            lidy.make_parser(
                lidy.File(name="<schema.yaml>", content=schema),
                {},
            )
        except Exception as e:
            error = e

    if outcome == "accept":
        if errorContains:
            s.abort(
                "'errorContains' cannot be specified when the expected outcome is 'accept'"
            )
        if error:
            s.fail(f"error: {error}")
    else:
        if error is None:
            s.fail("no error was found")
        if errorContains:
            if errorContains not in str(err):
                s.fail(f"error message '{err}' does not contain '{errorContains}'")

File: js/src/lidy_test.ts
import { readFileSync } from "fs"
import * as path from "path"
import * as specimen from "specimen-test"
import * as lidy from "."

var interpolationRegex = /\$\{([a-zA-Z0-9_]+)\}/g

function templateReadEntry(
  input: Record<string, string>,
  key: string,
): string | undefined {
  let value = input[key]
  let templateValue = input[`${key}Template`]
  if (value !== undefined && templateValue !== undefined) {
    throw new Error(
      `Found both '${key}' and '${key}Template'. Only on must be specified.`,
    )
  }
  if (value === undefined && templateValue === undefined) {
    return undefined
  }
  if (value !== undefined) {
    return value
  }

  return templateValue.replace(interpolationRegex, (match) => {
    let name = match.slice(2, -1)
    let value = input[name]
    if (value === undefined) {
      throw new Error(
        `the template interpolation key '${name} was not found in the input`,
      )
    }
    return value
  })
}

function specimenHandler(
  s: specimen.SpecimenContext,
  input: Record<string, string>,
) {
  // Target
  const box = input.box ?? "content"

  // Unpacking the input data

  // Text
  const text = templateReadEntry(input, "text")
  if (text === undefined) {
    s.fail("The 'text' entry is required")
  }

  // Expression and Schema
  const expression = templateReadEntry(input, "expression")
  let schema = templateReadEntry(input, "schema")

  if (box === "content") {
    if (expression !== undefined && schema !== undefined) {
      s.fail("'expression' and 'schema' cannot be specified together")
    }
    if (expression === undefined && schema === undefined) {
      s.fail("one of 'expression' and 'schema' must be specified")
    }
    if (expression !== undefined) {
      schema = `main:\n  ${expression.replace(/\n/g, "\n  ")}`
    }
  } else {
    let word = ""
    if (expression) {
      word = "expression"
    } else if (schema) {
      word = "schema"
    }
    if (word !== "") {
      s.fail(`box ${box} should not receive any ${word}`)
    }
  }

  // Name and Outcome
  const name = input.name
  let outcome = ""
  if (name.startsWith("accept")) {
    outcome = "accept"
  } else if (name.startsWith("reject")) {
    outcome = "reject"
  }
  if (outcome === "") {
    s.fail(
      `the 'name' entry should begin by 'accept' or 'reject', but it is: ${name}`,
    )
  }

  // // // // // // // // // // // // // // // // // // // //
  let error: Error | undefined
  let parser: lidy.Parser | undefined
  if (box === "content") {
    if (!schema) {
      s.fail("the schema cannot be empty")
    }
    try {
      parser = lidy.makeParser({ name: "<schema>.yaml", content: schema }, {})
    } catch (e) {
      let message = `error in schema: ${e.message}`
      s.abort(message)
    }
    try {
      parser.parse({ name: "<content>.yaml", content: text })
    } catch (e) {
      error = e
    }
  } else {
    schema = {
      lidySchemaExpression: () => `main:\n  ${text.replace(/\n/g, "\n  ")}`,
      lidySchemaDocument: () => text,
      lidySchemaRegexChecker: () => `main:\n  _regex: '${text}'`,
    }[box]?.()
    if (schema === undefined) {
      s.fail(`unknown test box: ${box}`)
    }
    try {
      lidy.makeParser({ name: "<schema>.yaml", content: schema }, {})
    } catch (e) {
      error = e
    }
  }

  // // // // // // // // // // // // // // // // // // // //

  const errorContains = input.errorContains
  if (outcome === "accept") {
    if (errorContains !== undefined) {
      s.fail(
        "'errorContains' cannot be specified when the expected outcome is 'accept'",
      )
    }
    if (error !== undefined) {
      s.fail(`error: ${error}`)
    }
  } else {
    if (error === undefined) {
      s.fail(`no error was found (${input.name})`)
    }
    if (errorContains !== undefined) {
      if (!error.message.includes(errorContains)) {
        s.fail(
          `error message '${error.message}' does not contain '${errorContains}'`,
        )
      }
    }
  }
}

function readLocalFile(filepath: string): specimen.File {
  const file = path.join(__dirname, filepath)
  return {
    path: file,
    content: readFileSync(file, "utf-8"),
  }
}

specimen.run(specimenHandler, [
  readLocalFile("../../testdata/collection/listOf.spec.yaml"),
  readLocalFile("../../testdata/collection/map.spec.yaml"),
  readLocalFile("../../testdata/collection/mapOf.spec.yaml"),
  readLocalFile("../../testdata/collection/merge.spec.yaml"),
  readLocalFile("../../testdata/collection/min_max_nb.spec.yaml"),
  readLocalFile("../../testdata/collection/tuple.spec.yaml"),
  readLocalFile("../../testdata/combinator/oneOf.spec.yaml"),
  readLocalFile("../../testdata/scalar/in.spec.yaml"),
  readLocalFile("../../testdata/scalar/range.spec.yaml"),
  readLocalFile("../../testdata/scalar/regexp.spec.yaml"),
  readLocalFile("../../testdata/scalarType/scalar.spec.yaml"),
  readLocalFile("../../testdata/schema/document.spec.yaml"),
  readLocalFile("../../testdata/schema/expression.spec.yaml"),
  readLocalFile("../../testdata/schema/mergeChecker.spec.yaml"),
  readLocalFile("../../testdata/schema/regex.spec.yaml"),
  readLocalFile("../../testdata/yaml/yaml.spec.yaml"),
])
